### PR TITLE
Change the default IO engine to pyogrio

### PR DESCRIPTION
**Pull request by martinfleis**
Still need to figure out the minimal pin, 0.4.0 may not be it.

Xref #3201

### CODE REVIEW

File: geopandas/io/tests/test_file.py
Comment (martinfleis) [inline geopandas/io/tests/test_file.py:282]: This is skipped because:

Older pandas (1.4.4, 1.5.3) with older pyogrio (0.5.0, 0.5.1) raise `pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 9999-12-31 00:00:00`

New pandas (2.2.1) in combination with old pyogrio (0.5.3) raise `AssertionError: assert dtype('<M8[ms]') == 'object'`.

And pandas 1.5.3 with recent pyogrio lead to `UnboundLocalError: local variable 'res' referenced before assignment` which has been fixed on pyogrio main in https://github.com/geopandas/pyogrio/pull/370

**Comment by jorisvandenbossche** [comment]
> Still need to figure out the minimal pin, 0.4.0 may not be it.

I think we can definitely go for a newer version, especially for a geopandas 1.0 where it is a new dependency anyway (0.4.0 is almost two years old, i.e. older than shapely 2.0). 

If we go for 1 year old, it's somewhere between 0.5 and 0.6. But I would personally even consider going for 0.7

File: doc/source/index.rst
Comment (jorisvandenbossche) [inline doc/source/index.rst:14]:
```suggestion
.. _pyogrio: https://pyogrio.readthedocs.io
```

**Comment by martinfleis** [comment]
> If we go for 1 year old, it's somewhere between 0.5 and 0.6. But I would personally even consider going for 0.7

I went with 0.5.0 in the end, which seems to work fine in our CI. 0.5.1 is what comes from anaconda defaults channel at the moment so it may be good to try and support it if there's no strict reason not to.

File: geopandas/io/tests/test_file.py
Comment (m-richards) [reply geopandas/io/tests/test_file.py:282]: I suppose technically for the 
> New pandas (2.2.1) in combination with old pyogrio (0.5.3) raise AssertionError: assert dtype('<M8[ms]') == 'object'.

case we could update the expectation to be for a ms resolution datetime. From what you're saying it seems that new pandas and new pyogrio passes the test as is? That's perhaps an oversight, there I would think it should give a lower resolution datetime to accomodate (especially since GDAL doesn't suport nanosecond precision anyway)

File: geopandas/tools/_show_versions.py
Comment (m-richards) [inline geopandas/tools/_show_versions.py:73]: In the latest version, you set the min version of pyogrio to 0.5, so `gdal_dir` should always be assignable in the try block above on line 64 (if pyogrio is available)

File: ci/envs/311-dev.yaml
Comment (m-richards) [inline ci/envs/311-dev.yaml:10]: does adding gdal here do anything, given that pyogrio is installed with pip? I would have thought the GDAL bundled in the wheel would be used regardless of whether we install with conda (but my understanding might be wrong)

**Review by m-richards — APPROVED** [review]
Thanks Martin, I think this looks good to me.

File: ci/envs/311-dev.yaml
Comment (martinfleis) [reply ci/envs/311-dev.yaml:10]: It is installed with pip but from source, not from a wheel. We need to compile it on installation for which we need GDAL available.

File: geopandas/io/tests/test_file.py
Comment (martinfleis) [reply geopandas/io/tests/test_file.py:282]: > we could update the expectation to be for a ms resolution datetime 

not sure if it is worth the effort of figuring out which combinations of versions require this.

> From what you're saying it seems that new pandas and new pyogrio passes the test as is? That's perhaps an oversight, there I would think it should give a lower resolution datetime to accomodate (especially since GDAL doesn't suport nanosecond precision anyway)

Yes, latest pandas with latest pyogrio pass fine. Which I believe is according to the expectations of this test, which checks that the out of bounds datetime is not a datetime object but a string and the resulting array dtype is object.

File: doc/source/community/ecosystem.md
Comment (brendan-ward) [inline doc/source/community/ecosystem.md:7]:
```suggestion
[`GEOS`](https://trac.osgeo.org/geos) library, `pyogrio`, allowing us to read and write
```

File: geopandas/io/file.py
Comment (brendan-ward) [inline geopandas/io/file.py:215]: The default should now by `pyogrio` here (and similar, below), right?

File: geopandas/io/tests/test_file.py
Comment (brendan-ward) [inline geopandas/io/tests/test_file.py:370]: Here and below, should this skip be moved into the test body, depending on value of `engine`, so that Fiona is still tested?

**Review by brendan-ward — COMMENTED** [review]
Thanks for working on this, and apologies for the slow response!  Just a quick review for now.

My preference would be to require pyogrio >= 0.7, as there were a few fairly significant API changes and bugfixes up to that version.  Otherwise, we risk exposing known issues based on installs with 0.5-0.6 that have already been fixed.

**Comment by martinfleis** [comment]
Bumped minimal version to 0.7.0. I think this will not resolve with defaults. Will wait for CI and install pyogrio from PyPI if needed.

File: geopandas/io/tests/test_file_geom_types_drivers.py
Comment (martinfleis) [inline geopandas/io/tests/test_file_geom_types_drivers.py:320]: To get pyogrio >= 0.7 in all our envs, we need to occasionally install it from PyPI. But it seems that GDAL in those wheels does not have spatialite 

```py
ValueError: unrecognized option 'SPATIALITE' for driver 'SQLite'
```

This is a bit verbose but I couldn't find a better solution as there does not seem to be a way of checking if spatialite is available or not. This way it gets checked if spatialite is available (conda) and xfails if it is not.

File: ci/envs/312-latest-conda-forge_no_pyproj.yaml
Comment (martinfleis) [reply ci/envs/312-latest-conda-forge_no_pyproj.yaml:9]: Yes

Comment (jorisvandenbossche) [reply ci/envs/312-latest-conda-forge_no_pyproj.yaml:9]: Does it actually more than fiona? I think for pyogrio, pyproj is only an optional dependency that is used in case of geopandas input for writing, and similarly for fiona in our `_to_file_fiona` we raise an error if pyproj is not installed?

Comment (jorisvandenbossche) [reply ci/envs/312-latest-conda-forge_no_pyproj.yaml:9]: Ah, but I assume our tests still pass with fiona because if pyproj is not available the `crs` will always be None, while in pyogrio we always import pyproj in the geopandas code path, regardless of whether `crs` is set.

Comment (martinfleis) [reply ci/envs/312-latest-conda-forge_no_pyproj.yaml:9]: Yes. To have a working env without pyproj, we need IO engine that does not require it, so Fiona. With pyogrio, we can't use IO without pyproj. I don't think that is an issue as the idea behind pyproj being optional is to depend on pandas+shapely only. In any case, this env needs to have Fiona and not pyogrio.

File: geopandas/io/tests/test_file_geom_types_drivers.py
Comment (brendan-ward) [reply geopandas/io/tests/test_file_geom_types_drivers.py:320]: Just so I'm tracking:
- does GDAL from conda-forge always have SpatialLite?
- do the Fiona wheels from PyPI always include SpatialLite?

Per [pyogrio #330](https://github.com/geopandas/pyogrio/issues/330), It looks like SpatialLite will be available in pyogrio >= 0.8 (not yet released).  So a version guard against pyogrio if installed via pip may be helpful here.

In case it helps, [here](https://github.com/geopandas/pyogrio/blob/main/pyogrio/tests/test_geopandas_io.py#L55) is how we are detecting if SpatialLite is available, by running an SQL query against the data source.

Comment (jorisvandenbossche) [reply geopandas/io/tests/test_file_geom_types_drivers.py:320]: > Per [pyogrio #330](https://github.com/geopandas/pyogrio/issues/330), It looks like SpatialLite will be available in pyogrio >= 0.8 (not yet released).


That issue mentions the GDAL small docker images that will now contain a GDAL built with libspatialite, but we are not using those docker images to build our wheels, so that shouldn't matter for the wheels.


> In case it helps, [here](https://github.com/geopandas/pyogrio/blob/main/pyogrio/tests/test_geopandas_io.py#L55) is how we are detecting if SpatialLite is available, by running an SQL query against the data source.

It seems we eventually didn't end up using that in pyogrio itself. The PR where this is being discussed eventually switched to use `INDIRECT_SQL` instead of skipping based on the presence of libspatialite (https://github.com/geopandas/pyogrio/pull/70#discussion_r856874188):

> If gdal is not built with spatialite support in SQLite,
you can use ``sql_dialect="INDIRECT_SQLITE"`` to be able to use spatialite 
functions on native SQLite files like Geopackage. 

---

For our wheels, we indeed don't enable libspatialite in the vcpkg GDAL build. Looking at it again, it seems we _could_ do that . @brendan-ward you commented about that in https://github.com/geopandas/pyogrio/issues/86#issuecomment-1109012589:

> Not sure if this matters, but noting for the record that our vcpkg build does not include `libspatialite` (we don't enable that as an add-on feature); it is LGPL licensed, which likely has license implications for the wheels we create.  I believe Fiona also does not build against `libspatialite`.

In https://github.com/geopandas/pyogrio/pull/161 I first added `libspatialite` feature, but then removed it again before the PR was merged, and I don't directly see any discussion or explanation that explains why I did that .. But there is a bit of discussion later at https://github.com/geopandas/pyogrio/pull/161#issuecomment-1269044332, which again mentions that fiona also doesn't have this.

Anyway, this is a discussion for pyogrio to have to see if we want to include libspatialite support in the wheels. 

Back to this PR, I am actually wondering why we explicitly use `spatialite=True` in the tests instead of using the default sqlite driver options

Comment (martinfleis) [reply geopandas/io/tests/test_file_geom_types_drivers.py:320]: > I am actually wondering why we explicitly use spatialite=True in the tests instead of using the default sqlite driver options

@theroggy do you remember why use explicitly used that when writing the test?

Comment (jorisvandenbossche) [reply geopandas/io/tests/test_file_geom_types_drivers.py:320]: Ah, it was added in https://github.com/geopandas/geopandas/pull/2829 just to explicitly test the spatialite option. I think this change is fine then

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Needs a merge of main

**Comment by kylebarron** [comment]
Now that pyogrio 0.8 has been released, are you planning to update the write side to default to pyogrio as well?

**Comment by martinfleis** [comment]
@kylebarron it should already default to pyogrio for writing. Or have I messed this PR?

**Comment by theroggy** [comment]
I suppose @kylebarron is referring to the new flavour of writing that uses the arrow interface of GDAL introduced in pyogrio 0.8?

Seems to soon though to switch to it as default. It is a very new feature in GDAL as well as in pyogrio that didn't have any serious real-life testing yet in either project...

**Comment by theroggy** [comment]
> Indeed, I conflated "pyogrio engine" with "arrow interface". So on both the read side and the write side, users will need to pass `use_arrow=True` to use the arrow interface, right? This is unfortunate (at least on the read side where it's better tested) but unavoidable unless pyarrow becomes a required dependency of pandas, I suppose.

Yes, for the read side I definitely agree it is unfortunate, but it is indeed to avoid pyarrow becoming a ~required dependency for geopandas...

**Comment by martinfleis** [comment]
We've been discussing whether we should use arrow automatically if pyarrow is installed in a similar way we used pygeos over shapely before. But until we know that there is no difference between the resulting files and the way they are read, it is a bit trickier and can be confusing for users. The chance that pyarrow is in some of your envs and not in the other without you knowing is much higher these days than it was with pygeos back then. So it is a recipe for surprises.

**Comment by m-richards** [comment]
>We've been discussing whether we should use arrow automatically if pyarrow is installed in a similar way we used pygeos over shapely before

I wonder if to avoid potential confusion but still be convenient we could expose a toplevel `geopandas.options` flag + environment variable. Or let the existing `geopandas.options.io_engine` accept `pyogrio_arrow` as an option, which would feed through use_arrow on read and write?

**Pull request merged**
